### ISSUE TITLE
nl床は physics でなく dispatch-latency 律速の可能性: graph capture 未検証 + 安い判定法

### ISSUE THREAD

**Issue by penta2himajin** [penta2himajin/qwisp#3]
nl(novel/high-entropy) の ~20 tok/s 床について、`notes/01-speedup-investigation.md` §2 は「batch=1 の原理床（MLX/Apple Silicon の物理）」と結論づけている。これを独立に裏取り（コード＋Web）した結果、**床は physics でなく dispatch/launch-latency 律速**であり、未検証の標準治療（command-graph capture）が1つ残っている、という指摘。実装前に二値で確定できる安い検証も併記する。

## 1. 数値: 帯域物理でなく launch 律速

- 3B-active・4bit ≈ **1.5GB/token**。dev 機（Max 級 ~400GB/s）の**メモリ帯域床 ≈ 3.75ms ＝ 理論 ~260 tok/s**。
- 実測 forward 固定費 **50ms（= 20 tok/s）はその ~13倍上**。帯域物理なら 260 出るはずが 20 ＝ **forward は帯域でなく dispatch/launch-latency 律速**（40 層の逐次カーネル投入の間 GPU がアイドル）。
- 弱機でもギャップは残る: M1/Neo（60–68GB/s）の帯域床 ≈ 22–25ms ＝ ~40–45 tok/s。50ms はなお ~2x 上。

## 2. 抜けている lever: command-graph capture（CUDA Graphs 相当）

`notes/01 §2` が否定した lever は **intra-kernel（fusion / mx.compile / 量子化）と token speculation のみ**。launch 律速 decode の de facto 治療である **graph capture が未検証**（コードにも無い。grep で出るのは routing capture と `asyncEval` のみ＝別物）。

文献裏取り:
- LLM decode は多数の短命カーネルで、1個ずつ CPU から launch すると host 側 launch overhead が end-to-end の有意割合になる。CUDA Graphs はグラフ全体を一度に launch して緩和、無効化で decode 性能が明確劣化（arXiv:2604.06664）。
- 極端な実例: RNN-T decode は launch 間 5µs に対しカーネル実行 1µs ＝ **GPU 80% idle、graph 化で 5x**（arXivः2406.03791）。Qwisp の 50ms 床と同型。
- 量産系でも CUDA Graph replay が launch＋同期の **~28% overhead を除去**（arXiv:2603.12707）。

→ 「50ms = 40 層 × ~1.2ms launch chain」は Apple Silicon 物理でなく **CPU→GPU 投入レイテンシ**。理論上は帯域床まで回収余地。

## 3. ただし catch（penta の実務判断＝「今は追わない」自体は妥当）

- **MLX に CUDA Graphs 相当が無い**。MLX の "graph" は lazy-eval の op fusion（memory roundtrip 削減）で、command-buffer を capture/replay する機構とは別。mx.compile はこの fusion ＝ matmul 支配層では net-negative（実証済）。→ 実現には **raw Metal の command-buffer capture（MTLIndirectCommandBuffer 等）を MLX の下で**組む必要があり、framework と戦う。
- **MoE 動的 routing が capture を阻害**。CUDA Graphs は静的 sequence 向けで per-token の expert 選択を capture できず、KTransformers 等が MoE 拡張を要した段階（arXiv:2603.22774）。resident なら indices を入力にした固定構造で回避可だが非自明。
- **弱機での旨味は ~2x に縮む**（帯域床が近いため。dev 機の 13x ギャップとは違う）。Apple Metal は同 VRAM の CUDA 比 1.1–4.2x 高レイテンシで dispatch tooling が未成熟（arXiv:2508.08531）。
- **speculation が nl を救えないのは正しい**（予測不能は投機不可）。量子化・fusion 否定も文献と整合。

## 4. 結論

- **正しい**: speculation は nl を救えない／量子化・fusion・compile は効かない。
- **言い過ぎ**: 「fundamental physics」。実体は **dispatch-latency 床**で、既知治療（graph capture, 他所で 5x/28% 実証）が未検証。ただし MLX に native 機構が無く、MoE routing と相まって高難度、弱機での上限 ~2x。
- ∴ **「今は追わない（nl は ~10%）」は妥当**だが、framing を **「MLX/Metal の tooling 限界」**に修正すべき。将来 nl を速くするなら、更なる fusion/quant でなく **graph-capture-below-MLX が唯一の筋**。

## 5. 投資前の安い検証（推奨・最優先）★

graph capture を組む前に、**50ms が CPU-submit 律速か GPU-exec 律速かを二値判定**する。RNN-T の「GPU 80% idle」測定の型。

- **計測**: `forward-cost` probe（hot-pin で IO 排除済）に **GPU active 時間（Metal GPU timestamp / `MTLCommandBuffer.GPUStartTime`–`GPUEndTime`、または Metal System Trace の GPU 占有率）vs wall-clock** を追加。1 forward あたりの GPU-busy ms と wall ms を出すだけ。
- **判定**:
  - GPU idle 率が高い（busy ≪ wall）→ **dispatch 律速確定** → graph capture に価値（弱機で ~2x の上限つき、要 framework 工事）。
  - GPU がほぼ busy（busy ≈ wall）→ 実質 **GPU-exec 床** → penta の「床」が正しく graph capture も無駄。
- **コスト**: probe に timestamp を足すだけ（モデル不要 or 既存 stream 経路）。半日。これで「physics 床か tooling 床か」が確定し、深掘り判断ができる。

## 補足: 部分的な軽い手（graph capture 未満）
- **層内カーネル数の削減**（GDN+MoE をより少数の dispatch に融合）。ただし `fuseGDN` は batch=1 で ~2% 実測＝層内融合の天井は低い（逐次は inter-layer chain ゆえ）。graph capture（inter-layer）が本丸。
- これらは §5 の判定が「dispatch 律速」と出た場合のみ意味を持つ。

## 参照
- CUDA Graphs for decode launch overhead: arXiv:2604.06664 / 2406.03791(RNN-T 5x) / 2603.12707(~28%)
- MoE dynamic routing と graph capture の壁: arXiv:2603.22774(KTransformers 言及)
- Apple Silicon decode latency vs CUDA: arXiv:2508.08531
- MLX = lazy fusion（CUDA Graphs 非等価）: 一般解説（compute-market 2026 ほか）
- 前提: notes/01-speedup-investigation.md §2 / notes/02-device-matrix.md
- ※ arXiv ID は引用前に再確認推奨

**Comment by penta2himajin** [comment]
## §5 検証完了: dispatch/sync 律速と実測確定（GPU ~35% busy, M1 Max）

branch `issue-3-dispatch-vs-exec`（commit 146da8c）で §5 の二値判定を実装し、両法で裏取りした。

### 前提の訂正
issue が想定した「probe に `MTLCommandBuffer.GPUStartTime–GPUEndTime` を足すだけ（半日）」は **実際には使えない**。mlx-swift は command buffer を内部管理して公開していないため、その timestamp を Swift から取れない。代わりに 2 法で測定:

### 1. 自動 probe（`QWISP_RUN=forward-gpu-busy`, mlx 再ビルド不要）
- **build-only(eval せず forwardHidden のみ) ≈37ms ≈ K=1 eval 35ms** → forward は lazy でなく **内部で per-layer routing 同期を強制**（router logits を materialize して expert gather する round-trip）。
- **CPU-busy ≈ 0.51 cores**（`getrusage` の process CPU 時間 / wall）= wall の約半分は CPU が encode/sync でビジー、残り半分は GPU 待ちで block。
- 独立 forward を K 本まとめて 1 eval する pipeline 法は **flat（0.94x）** = 同期が forward 内で逐次化し、末尾 eval をまとめても GPU idle を埋められない。

```
K=1:  35.35 ms/forward  (CPU-busy=0.51 cores)
K=2:  33.76 ms/forward  (CPU-busy=0.52 cores)
K=4:  33.30 ms/forward  (CPU-busy=0.53 cores)
K=8:  33.08 ms/forward  (CPU-busy=0.53 cores)
```

### 2. Metal System Trace（M1 Max, ground truth）
`xcrun xctrace record --template 'Metal System Trace'` で forward ループを記録し、`metal-gpu-intervals`（GPU 実行区間）を process=qwisp・depth=0 で union 算出（`tools/gpu_busy_from_trace.py`）:

- **GPU-busy ≈ 34.6%**（steady median 33.8% / p90 51%）= **GPU は wall の ~65% アイドル**

### 結論
- **床は GPU-exec（compute/帯域）でなく dispatch/sync-latency 律速で確定**。理論帯域床 ~260 tok/s に対し 20 が出る真因は per-layer の CPU↔GPU routing round-trip（GPU 65% idle）。→ issue の「§4: fundamental physics は言い過ぎ、実体は tooling/sync 床」を **実測で支持**。
- **ただし idle の正体が MoE 動的 routing の round-trip** ゆえ、§3 の指摘どおり command-graph capture は per-token の expert 選択を capture できず native には効かない。効く lever は intra-kernel fusion（既存 fuseGDN は batch=1 で ~2%）でなく **sync/round-trip 除去**（routing 予測で expert 選択を先回りし層跨ぎ materialize を消す）。これは別途高難度（予測精度 ~82-84% 頭打ち実績）で、弱機（M1/Neo 帯域床 ~40-45 tok/s）では旨味 ~2x に縮む。
- ∴ **「nl は今追わない」判断は妥当**。framing のみ「MLX/Metal の sync-tooling 限界」に修正（notes/01 §2-4 に記録）。

### 再現
```sh
# 自動 probe
QWISP_RUN=forward-gpu-busy QWISP_CACHE_C=64 QWISP_FC_REPS=30 <qwisp-poc> stream
# GPU 占有率（Metal System Trace → union 算出）
xcrun xctrace record --template 'Metal System Trace' --output /tmp/qwisp_gpu.trace \
  --env QWISP_RUN=forward-gpu-busy --env QWISP_GPUBUSY_K=1 --env QWISP_FC_REPS=200 \
  --env QWISP_MODEL=... --env QWISP_MTP_REF=/tmp/qwisp_mtp_ref.safetensors \
  --launch -- <qwisp-poc> stream
xcrun xctrace export --input /tmp/qwisp_gpu.trace \
  --xpath '/trace-toc/run[@number="1"]/data/table[@schema="metal-gpu-intervals"]' \
  --output /tmp/qwisp_gpuint.xml
python3 tools/gpu_busy_from_trace.py /tmp/qwisp_gpuint.xml
```

**Comment by penta2himajin** [comment]
## Resolved & Close: 床=dispatch/sync 律速を実測確定、lever は PR #4 で ship（C=256 のみ）

### §5 検証（実測確定）
- 「probe に `GPUStartTime–GPUEndTime` を足すだけ」は **mlx-swift が command buffer 非公開で不可**。Metal System Trace ＋ `getrusage` で代替（branch `issue-3-dispatch-vs-exec`）。
- **forward GPU-busy ≈34.6%（~65% idle）**、CPU-busy≈0.51 cores、build-only≈eval。→ **床は GPU-exec/帯域でなく dispatch/sync-latency 律速**（per-layer MoE routing round-trip）。本 issue §4「fundamental physics は言い過ぎ／tooling-sync 床」を**実測で支持**。

### lever 実装（PR #4, merged）
- **round-trip 除去（no-sync）は全 expert resident なら bit-exact**（`max|Δhidden|=0`）。forward 31.9→16.1ms=**1.99x**、greedy decode **C=256 32.7→56.7 tok/s=1.73x, lossless**。`runSuffixSpec` に C≥256 auto no-sync 配線（mix spec 224 vs 196、無回帰）。
- escalation（C<256 を exact 化）は cold-start で net-negative。8GB 拡張（cross-layer 予測 pipeline）は 64/64 bit-exact だが **0.26–0.46x（予測が round-trip を*追加*: 80+ sync/token > 40）= NO-GO**。

### 確定した characterization（再燃防止）
> **per-layer sync は streaming を駆動するために存在する**（CPU が「どの expert が欠けてるか」を知らないと SSD fetch できない）。
> - 全常駐(C=256) → streaming 不要 → CPU round-trip 不要 → no-sync GPU gather → **1.73x bit-exact**。
> - 部分常駐(C<256) → streaming に sync 必須 → 床が残る。

∴ 制約デバイス(8GB/16GB)の nl 床は **Apple Silicon 物理でなく streaming 要件に内在する床（容量壁）**。「physics ではない」と「streaming する限り fundamental」は両立。**streaming 能力 ⟺ sync 床はトレードオフ**で、同一 forward で「bit-exact 高速」と「streaming」は両立しない。

### 含意
- **24GB+ 全常駐 tier は無料で 1.73x（bit-exact）** → device matrix 上 C=256 tier が streaming tier から speed でも分離。
- **8GB/Neo の nl は床のまま**だが ~10% の workload で、構造的に確定（40 個の逐次 routing sync が不可避）。
- 将来「round-trip 消せないか」が再燃したら → **全常駐以外は道無し（PR #4 で実証）**で即決着。

→ 床の正体・lever・限界が確定。close（completed）。記録は notes/01 §2-4/2-5/2-6。

**Comment by penta2himajin** [comment]
## 訂正: no-sync 1.73x の対象 tier は「24GB+」でなく **32GB+（C256 full-resident）**

上のクローズコメントの「**24GB+ 全常駐 tier は無料で 1.73x**」は不正確。正しくは **32GB+**。

- no-sync は **C=256（full-resident, RSS 20.5GB）必須**（全 expert 常駐で GPU slot-table remap が bit-exact。miss が出ると sync 復活）。
- 確定 tier（notes/02）: **24GB→C192 / 32GB+→C256**。24GB は **C192（15.9GB, 75–81% 常駐, miss 数%）＝near-full ＝ streaming-with-sync** で、**no-sync の対象外**。
- 24GB に C256（20.5GB）は数値上載るが headroom ~3.5GB のみ → memory pressure で expert eviction → 実質 streaming に転落。**安全な full-resident tier は 32GB から**。

PR #4 本文の「C=256（~21GB, 24GB+機）」は「C256 を置くには最低 24GB 要る」の意で、「24GB が既定で C256 を走らせる」ではない。それを読み違えた。

**確定: no-sync 1.73x bit-exact の恩恵は 32GB+（C256）tier のみ。24GB は C192 で対象外。** canonical な tier 定義は notes/02 §3 が正。
